### PR TITLE
[CAS-1190 CAS-1191] Add higher-level UI components

### PR DESCRIPTION
**Pull request by bychkovdmitry**
https://stream-io.atlassian.net/browse/CAS-1190
https://stream-io.atlassian.net/browse/CAS-1191

### 🎯 Goal

Create self-contained components for the channel list screen and for the message list screen:
- ChannelListActivity
- ChannelListFragment
- MessageListActivity
- MessageListFragment

### 🛠 Implementation details

The implementation details are described in the [PoC](https://docs.google.com/document/d/112s-OGzzJrr272eBKOjVamIxD0FYzREz9iAprA5zgBg/edit?usp=sharing)

### 🎨 UI Changes

| ChannelListAcitivty | MessageListActivity |
| --- | --- |
 | photo_2021-08-16 18 18 05| photo_2021-08-16 18 18 07 |

### 🧪 Testing

Start the corresponding activity:

```kotlin
context.startActivity(ChannelListActivity.createIntent(context))
```

### ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] PR targets the `develop` branch
- [x] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS (task created))
- [x] Reviewers added

### 🎉 GIF

giphy (2)

### CODE REVIEW

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/ChannelListActivity.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * Creates an instance of [ChannelListFragment]. Override this method if you want to create an
+     * instance of [ChannelListFragment] with custom arguments or if you want to create a subclass
+     * of [ChannelListFragment].
+     */
+    protected open fun createChannelListFragment(): ChannelListFragment {
+        return ChannelListFragment.newInstance {
+            setFragment(ChannelListFragment())
+            setTheme(R.style.StreamUiTheme_ChannelListScreen)
```
Comment (bychkovdmitry) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/ChannelListActivity.kt:37]: This theme is only necessary to: 
- hide "View info" button in the channel options dialog
- hide action button in the channel list header  
Maybe better to make then hidden by default but it is a braking change

**Review by adasiewiczr — APPROVED** [review]
Good job! Let's wait for more reviews

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/MessageListFragment.kt [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        setupMessageList()
+        setupMessageInput()
+    }
+
+    @StyleRes
+    protected open fun getTheme(): Int {
+        return themeResId
+    }
+
+    protected open fun setupMessageListHeader() {
```
Comment (zsmb13) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/MessageListFragment.kt:91]: Really great docs on a lot of the declarations in this PR - can we please add docs on all of these `open` functions too?

File: UNRELEASED_CHANGELOG.md [hunk 3]
```diff
@@ -71,7 +71,12 @@
 ### ⬆️ Improved
 
 ### ✅ Added
-
+- Added self-contained higher-level UI components:
```
Comment (zsmb13) [inline UNRELEASED_CHANGELOG.md:74]: Awesome work all around, very nicely done.

I have just one question: what about Docusaurus docs? 😀

**Review by zsmb13 — CHANGES_REQUESTED** [review]

File: UNRELEASED_CHANGELOG.md (hunk 3 above)
Comment (bychkovdmitry) [reply UNRELEASED_CHANGELOG.md:74]: Was not able to complete the docs in the previous sprint so we have a task to do that in the current sprint

**Review by filbabic — APPROVED** [review]
Looks awesome - mainly the only thing I'd love for us to add here is docs for all of the functions in our components and their setup, state production etc.

Other than that - looks awesome!

File: stream-chat-android-docs/src/main/kotlin/io/getstream/chat/docs/kotlin/Screens.kt [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                    showHeader(false)
+                }
+            }
+
+            override fun onUserAvatarClick() {
+                // Handle avatar click
+            }
+        }
+
+        class CustomChannelListFragment : ChannelListFragment() {
```
[the remark this answers is not in the document]
Comment (zsmb13) [reply stream-chat-android-docs/src/main/kotlin/io/getstream/chat/docs/kotlin/Screens.kt:48]: This is in the docs module 😉 We use this to store snippets that are in our CMS / other documentation. This way we realize if we break them and keep them updated.

**Review by zsmb13 — APPROVED** [review]
LGTM

**Pull request merged**